src/services/websocket.rs

```diff
@@ -59,33 +73,82 @@ impl WebSocketService {
             return Err("Failed to created websocket with given URL");
         }
 
-        let ws = ws.unwrap();
-        ws.set_binary_type(SocketBinaryType::ArrayBuffer);
+        let ws = ws.map_err(|_| "failed to build websocket")?;
+        cfg_match! {
+            feature = "std_web" => ws.set_binary_type(SocketBinaryType::ArrayBuffer),
+            feature = "web_sys" => ws.set_binary_type(BinaryType::Arraybuffer),
+        };
         let notify = notification.clone();
-        ws.add_event_listener(move |_: SocketOpenEvent| {
-            notify.emit(WebSocketStatus::Opened);
-        });
+        let listener_open =
+            move |#[cfg(feature = "std_web")] _: SocketOpenEvent,
+                  #[cfg(feature = "web_sys")] _: &Event| {
+                notify.emit(WebSocketStatus::Opened);
+            };
         let notify = notification.clone();
-        ws.add_event_listener(move |_: SocketCloseEvent| {
-            notify.emit(WebSocketStatus::Closed);
-        });
+        let listener_close =
+            move |#[cfg(feature = "std_web")] _: SocketCloseEvent,
+                  #[cfg(feature = "web_sys")] _: &Event| {
+                notify.emit(WebSocketStatus::Closed);
+            };
         let notify = notification.clone();
-        ws.add_event_listener(move |_: SocketErrorEvent| {
-            notify.emit(WebSocketStatus::Error);
-        });
-        ws.add_event_listener(move |event: SocketMessageEvent| {
-            if let Some(bytes) = event.data().into_array_buffer() {
-                let bytes: Vec<u8> = bytes.into();
-                let data = Ok(bytes);
-                let out = OUT::from(data);
-                callback.emit(out);
-            } else if let Some(text) = event.data().into_text() {
-                let data = Ok(text);
-                let out = OUT::from(data);
-                callback.emit(out);
-            }
-        });
-        Ok(WebSocketTask { ws, notification })
+        let listener_error =
+            move |#[cfg(feature = "std_web")] _: SocketErrorEvent,
+                  #[cfg(feature = "web_sys")] _: &Event| {
+                notify.emit(WebSocketStatus::Error);
+            };
+        let listener_message =
+            move |#[cfg(feature = "std_web")] event: SocketMessageEvent,
+                  #[cfg(feature = "web_sys")] event: &Event| {
+                #[cfg(feature = "web_sys")]
+                let data = event.dyn_ref::<MessageEvent>().unwrap().data();
+                #[cfg(feature = "std_web")]
+                let text = event.data().into_text();
+                #[cfg(feature = "web_sys")]
+                let text = data.as_string();
+                #[cfg(feature = "std_web")]
+                let bytes = event.data().into_array_buffer();
+                #[cfg(feature = "web_sys")]
+                let bytes = Some(data);
```

Comment:
`cfg_match`'s would be nice here

Reply:
Done, I seem to have overlooked this part.